.github/workflows/molecule-actions-core-dhcp_server.yml

```diff
@@ -0,0 +1,39 @@
+---
+name: core/dhcp_server unit testing
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - 'roles/core/dhcp_server/**'
+      - '.github/workflows/molecule-actions-core-dhcp_server.yml'
+  pull_request:
+    paths:
+      - 'roles/core/dhcp_server/**'
+      - '.github/workflows/molecule-actions-core-dhcp_server.yml'
+
+jobs:
+  molecule_test:
+    name: ${{ matrix.distro }} - molecule test
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        distro:
+          - centos:8
```

Comment:
Is there any failure with CentOS 7?

Reply:
Yes it was't running in my platform. I was getting several errors in the container, unfortunately I was't able to solve this.
For now only tests in cento8 are working. Do you have any idea about this error? 

Errror output :

`TASK [dhcp_server : package █ Install dhcp] ************************************
fatal: [management1]: FAILED! => {"changed": false, "changes": {"installed": ["dhcp"]}, "msg": "http://mirrors.vcea.wsu.edu/centos/7.8.2003/os/x86_64/Packages/dhcp-4.2.5-79.el7.centos.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: mirrors.vcea.wsu.edu; Unknown error\"\nTrying other mirror.\nhttp://mirrors.vcea.wsu.edu/centos/7.8.2003/os/x86_64/Packages/dhcp-libs-4.2.5-79.el7.centos.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: mirrors.vcea.wsu.edu; Unknown error\"\nTrying other mirror.\nhttp://mirror.centos.lax1.serverforge.org/7.8.2003/updates/x86_64/Packages/bind-export-libs-9.11.4-16.P2.el7_8.6.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: mirror.centos.lax1.serverforge.org; Unknown error\"\nTrying other mirror.\nhttp://mirrors.vcea.wsu.edu/centos/7.8.2003/os/x86_64/Packages/dhcp-common-4.2.5-79.el7.centos.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: mirrors.vcea.wsu.edu; Unknown error\"\nTrying other mirror.\nhttp://repo1.sea.innoscale.net/centos/7.8.2003/os/x86_64/Packages/dhcp-libs-4.2.5-79.el7.centos.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: repo1.sea.innoscale.net; Unknown error\"\nTrying other mirror.\nhttp://centos.s.uw.edu/centos/7.8.2003/os/x86_64/Packages/dhcp-4.2.5-79.el7.centos.x86_64.rpm: [Errno 14] curl#6 - \"Could not resolve host: centos.s.uw.edu; Unknown error\"\nTrying other mirror 
[...] `

Reply:
I'm having a hard time trying to reproduce this on my dev system (Debian 10.6)... The IP subnet used for the Docker network `bbnet` conflicts with my Vagrant setup (which uses a copy of the example inventory for dev/testing of playbooks). It would be nice to use a different subnet than 10.10.0.0/16 for developers who run a virtual environment for dev & integration tests and molecule for unit tests, all on the same host.

With a different subnet, the container `management1` loses access to the network after the start of the firewalld service, the system cannot resolve any address, which seems close to what you observe. However, I observe this with CentOS 8 only. With CentOS 7, the task `[dhcp_server : package █ Install dhcp]` succeeds. 😵

Reply:
On my workstation, the CentOS 8 test still fails while CentOS 7 succeeds (after a tweak in verify.yml to check dhcp package for EL7).

```
$ MOLECULE_DISTRO=centos:8 molecule test
[...]
    TASK [dhcp_server : package █ Install dhcp-server] *****************************
fatal: [management1]: FAILED! => {"changed": false, "msg": "Failed to download packages: Curl error (6):
Couldn't resolve host name for http://mirrorlist.centos.org/?release=8&arch=x86_64&repo=BaseOS&infra=container
[Could not resolve host: mirrorlist.centos.org]", "results": []}

ERROR: 
An error occurred during the test sequence action: 'converge'. Cleaning up.
```

```
$ MOLECULE_DISTRO=centos:7 molecule test
[...]
$ echo $?
0
```

The use of firewalld in Docker containers is not standard per se, it looks like it triggers some bad behaviour when we define additional Docker networks. I understand you defined this additional network to be able to start the dhcp service with the expected IP address. I had the same issue with dns_server and eventually test the role with `start_service: false`.

Would it be possible to remove the network bbnet from this scenario? It would allow to test the role with CentOS 7 and 8 (both are supported) with the firewall enabled. We can still consider another scenario to do advanced test, without the firewalld service.

Note: I have some code sleeping in my computer for months for a new scenario for dns_server (now advanced_dns_server) that uses Docker networks the way you do here. I will refresh it and push to a new branch. We might converge to some good practice.

Reply:
BTW, you can test if the dhcpd.conf is correct by running `dhcpd -t -cf /etc/dhcp/dhcpd.conf` instead of starting the service. It can be a good test to add to the verify.yml.

Reply:
In a stack we made 2 years ago, there was some tests like that. It was pretty cool !

Reply:
I prefere to keep the firewall tests as-is because we need to cover/test most of the tasks of this role.
We do not need to run `dhcpd -t -cf /etc/dhcp/dhcpd.conf` if the `dhcpd `service is correctly started.
In the past I already had several problems creating a new docker network, it always gave me a headache !
I believed that the easiest way to manage the docker network is using environment variables before running `molecule test`.
We can discover the container ip using `docker inspect.` So basically we should have something like this (a little dirty but it should work): 

```
# cat .github/workflows/molecule-actions-core-dhcp_server.yml

      [...]
	  
      - name: Run molecule test for core/dhcp_server
        run: |
	  export IP_CONTAINER=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" management1)
	  export IP_PREFIX_SUBNET1=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" management1 | cut -d "." -f 1-2)
          cd roles/core/dhcp_server && molecule test
```

Finally the provisioner section in `molecule.yml` will look like:

```
# cat molecule/default/molecule.yml

[...]

provisioner:
  name: ansible
  inventory:
    host_vars:
      management1:

        network_interfaces:
          - interface: enp0s3
            ip4: ${IP_CONTAINER}
            mac: 08:00:27:dc:f8:f5
            network: ice1-1

        [...]

        networks:
          ice1-1:
            subnet: ${IP_PREFIX_SUBNET1}.0.0
            prefix: 16
            netmask: 255.255.0.0
            broadcast: ${IP_PREFIX_SUBNET1}.255.255
            dhcp_unknown_range: ${IP_PREFIX_SUBNET1}.254.1 ${IP_PREFIX_SUBNET1}.254.254
            gateway: ${IP_PREFIX_SUBNET1}.2.1
            is_in_dhcp: true
            is_in_dns: true
            services_ip:
              pxe_ip: ${IP_CONTAINER}
              dns_ip: ${IP_CONTAINER}
              repository_ip: ${IP_CONTAINER}
              authentication_ip: ${IP_CONTAINER}
              time_ip: ${IP_CONTAINER}
              log_ip: ${IP_CONTAINER}

    hosts:
      compute:
        hosts:
          compute1:
            j2_current_iceberg_network: ice1
            network_interfaces:
            - interface: enp0s3
              ip4: ${IP_PREFIX_SUBNET1}.3.1
              mac: 08:00:27:dc:f8:f6
              network: ice1-1
```

This approach allow us to start the services `firewalld `and `dhcpd` using the default docker network.
What do you think about it?

Reply:
> I prefere to keep the firewall tests as-is because we need to cover/test most of the tasks of this role.
> We do not need to run `dhcpd -t -cf /etc/dhcp/dhcpd.conf` if the `dhcpd `service is correctly started.

Yes, I also prefer to keep the firewall tests. The issue here seems to be between firewalld and the Docker network `bbnet`. The role needs this network for a single purpose: start the `dhcpd` service. If the role does not start the service but verify the configuration with `dhcpd -t -cf /etc/dhcp/dhcpd.conf`, we're all good.

> In the past I already had several problems creating a new docker network, it always gave me a headache !

Yeah, this was somewhat limited until molecule 3.0.5 release in July.

> I believed that the easiest way to manage the docker network is using environment variables before running `molecule test`.
> We can discover the container ip using `docker inspect.` So basically we should have something like this (a little dirty but it should work):
> [...]
> This approach allow us to start the services `firewalld `and `dhcpd` using the default docker network.
> What do you think about it?

Did you test this?
I believe the action will evaluate the docker inspect command before the start of molecule, when container management1 does not exist yet.

Reply:
It was only an idea but you are right the container does not exist yet, we need to create a dummy container before running `molecule test` , this container will be destroyed by the default scenario at the beginning of the molecule execution.
I am thinking in something like this, but I am not really convinced with this workaround: 

```
# cat .github/workflows/molecule-actions-core-dhcp_server.yml

      [...]
	  
      - name: Run molecule test for core/dhcp_server
        run: |
	  docker pull quay.io/actatux/ansible-centos:8
          docker create --name management1 quay.io/actatux/ansible-centos:8
          docker start management1
	  export IP_CONTAINER=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" management1)
	  export IP_PREFIX_SUBNET1=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" management1 | cut -d "." -f 1-2)
          cd roles/core/dhcp_server && molecule test
```
		  
I have tested it in my local environment and it works for centos7 and centos8.

For now I only see two solutions:

- either we test only in a centos8 container with the declarative docker network (my first proposition) 
- or we use the workaroud with environments vars to be able to test centos7 and centos8 containers (my second proposition) 

We need to decide soon which solution to use, actually I am stuck with the roles using networks (cf: dhcp, advanced-dhcp, pxe_stack, etc) due to this.

If you have another idea it will be more than welcome :)

Reply:
> either we test only in a centos8 container with the declarative docker network (my first proposition)

No, it's not reliable. CentOS 8 does not work on my workstation, CentOS 7 does.

> or we use the workaroud with environments vars to be able to test centos7 and centos8 containers (my second proposition)

That's definitely too hacky. We will quickly run in troubles when spawning several containers.

> If you have another idea it will be more than welcome :)

If you really want to stick with testing firewalld + start services that need a dedicated Docker network in the same molecule scenario, my idea is to find a way to make it work out-of-the-box, as it does with default network. Running firewalld inside a container is not a common use-case since its backend is running in-kernel, it's not a surprise to have strange behaviour. That being said, we might miss a tuning somewhere?

Why do you need the `bbnet` Docker network in this scenario?

Reply:
> Why do you need the bbnet Docker network in this scenario?


In order to run the dhcpd service I need to know the subnet/Gateway and the IP to configure the dhcp server and a fake compute node, that is why I am declaring the `bbnet` Docker network.
Perhaps I am going a little too far and as you said a simple `dhcpd -t -cf /etc/dhcp/dhcpd.conf ` could be enough as the more advanced unitary test.

Additionally more advanced test (including functional tests) could be performed using virtual machines in our internal platform. I mean every roles that need a specific netwotk will be covered in this part.

Reply:
The fake compute node is not a container, it does not need any Docker network. The ansible controller will load the inventory and create the dhcp configuration from the template.

Question is: Why do you need to run the dhcpd service?

If it's to test the configuration is correct, the `dhcpd -t ...` might be sufficient, no need to start the service.

This is the default scenario. We can then create more scenario without firewalld to add Docker networks and start some services. (The firewall configuration for the role being tested in the default scenario, we don't need to test it further).

Reply:
> Question is: Why do you need to run the dhcpd service?

Yes, I want to test if the configuration is correct via the service.
For now, I will rework this PR and the other one (advanced_dhcp_server) based on our discussion.
We will modify these tests in the future once we find a better solution concerning the docker network.
Thanks.